lib_test/protocol_test.ml

```diff
@@ -16,127 +16,214 @@ open OUnit
+          let t =
+            Client.negotiate channel "export1"
+            >>= fun (t, size, flags) ->
+            Lwt.return ()
+          in
+          Lwt_main.run t
+        )
+
+  let server_negotiation =
+    "Perform a negotiation using the second version of the protocol from the
+     server's side."
+    >:: fun () ->
+      with_server_channel v2_negotiation_start (fun channel ->
+          let t =
+            Server.connect channel ()
+            >>= fun (export_name, svr) ->
+            OUnit.assert_equal ~msg:"The server did not receive the correct export name" "export1" export_name;
+            Lwt.return_unit
+          in
+          Lwt_main.run t
+        )
+end
+
+module V2_list_export_disabled = struct
+
+  let v2_list_export_disabled = [
+    `Server, "NBDMAGIC"; (* read *)
+    `Server, "IHAVEOPT";
+    `Server, "\000\001"; (* handshake flags: NBD_FLAG_FIXED_NEWSTYLE *)
+    `Client, "\000\000\000\001"; (* client flags: NBD_FLAG_C_FIXED_NEWSTYLE *)
+    `Client, "IHAVEOPT";
+    `Client, "\000\000\000\003"; (* NBD_OPT_LIST *)
+    `Client, "\000\000\000\000";
+    `Server, option_reply_magic_number;
+    `Server, "\000\000\000\003";
+    `Server, "\128\000\000\002";
+    `Server, "\000\000\000\000";
+  ]
+
+  let client_list_disabled =
+    "Check that if we request a list of exports and are denied, the error is
+     reported properly."
+    >:: fun () ->
+      with_client_channel v2_list_export_disabled (fun channel ->
+          let t =
+            Client.list channel
+            >>= function
+            | `Error `Policy ->
+              Lwt.return ()
+            | _ -> failwith "Expected to receive a Policy error" in
+          Lwt_main.run t
+        )
+
+  let server_list_disabled =
+    "Check that the server denies listing the exports, and the error is
+     reported properly."
+    >:: fun () ->
+      with_server_channel v2_list_export_disabled (fun channel ->
+          let t () =
+            Server.connect channel ()
+            >>= fun (_export_name, _svr) ->
+            Lwt.return_unit
+          in
+          (* The Client.list function currently does not send NBD_OPT_ABORT when it
+             should, but incorrectly disconnects, so we expect this error from the
+             server side. *)
+          OUnit.assert_raises
+            Failed_to_read_empty_stream
+            (fun () -> Lwt_main.run (t ()))
+        )
+end
+
+module V2_list_export_success = struct
+  let v2_list_export_success = [
+    `Server, "NBDMAGIC"; (* read *)
+    `Server, "IHAVEOPT";
+    `Server, "\000\001"; (* handshake flags: NBD_FLAG_FIXED_NEWSTYLE *)
+    `Client, "\000\000\000\001"; (* client flags: NBD_FLAG_C_FIXED_NEWSTYLE *)
+    `Client, "IHAVEOPT";
+    `Client, "\000\000\000\003"; (* NBD_OPT_LIST *)
+    `Client, "\000\000\000\000";
+
+    `Server, option_reply_magic_number;
+    `Server, "\000\000\000\003";
+    `Server, "\000\000\000\002"; (* NBD_REP_SERVER *)
+    `Server, "\000\000\000\011";
```

Comment:
I think this is actually incorrect, this lines should be removed, that's why I didn't add a server test, because it failed. See https://github.com/xapi-project/nbd/issues/103